Mage/src/main/java/mage/abilities/effects/common/MayCastTargetThenExileEffect.java

```diff
@@ -15,44 +16,75 @@
 import mage.util.CardUtil;
 
 /**
- * @author xenohedron
+ * @author xenohedron, Susucr
  */
 public class MayCastTargetThenExileEffect extends OneShotEffect {
 
     private final Duration duration;
-    private final boolean noMana;
+
+    private final CardUtil.CastManaAdjustment manaAdjustment;
+
+    // Set to true with `withNoExile` if not wanting the exile clause (rare).
+    private boolean noExile;
```

Comment:
The exile part is a small part in this effect. The central part is the "You may cast the card".

I would rather rename the effect "MayCastTargetEffect" then add a boolean `doExileOnResolve`, rather than not having Tinybones with the other cards there.